UnityGLTF/Assets/UnityGLTF/Examples/OrbitCameraController.cs

```diff
@@ -44,38 +52,52 @@ void Start()
 
 		void LateUpdate()
 		{
-			if (target)
+			var height = Display.main.renderingHeight;
+			var width = Display.main.renderingWidth;
+
+			if (Input.GetMouseButton(0))
 			{
-				if (Input.GetMouseButton(0))
-				{
-					x += Input.GetAxis("Mouse X") * xSpeed * 0.06f;
-					y -= Input.GetAxis("Mouse Y") * ySpeed * 0.06f;
+				x += Input.GetAxis("Mouse X") * xSpeed * 0.06f;
+				y -= Input.GetAxis("Mouse Y") * ySpeed * 0.06f;
 
-					y = ClampAngle(y, yMinLimit, yMaxLimit);
+				y = ClampAngle(y, yMinLimit, yMaxLimit);
 
-					rotation = Quaternion.Euler(y, x, 0);
-				}
+				rotation = Quaternion.Euler(y, x, 0);
+			}
+			else if (Input.GetMouseButton(1))
+			{
+				var prevMouseWorldPosition = ProjectScreenPointToTargetPlane(prevMousePosition);
+				var mouseWorldPosition = ProjectScreenPointToTargetPlane(Input.mousePosition);
 
-				var height = Display.main.renderingHeight;
-				var width = Display.main.renderingWidth;
+				targetPosition += prevMouseWorldPosition - mouseWorldPosition;
+			}
 
-				var mouseOverRenderArea =
-					Input.mousePosition.x >= 0 &&
-					Input.mousePosition.x <= width &&
-					Input.mousePosition.y >= 0 &&
-					Input.mousePosition.y <= height;
+			var mouseOverRenderArea =
+				Input.mousePosition.x >= 0 &&
+				Input.mousePosition.x <= width &&
+				Input.mousePosition.y >= 0 &&
+				Input.mousePosition.y <= height;
 
-				if (Input.GetMouseButton(0) || mouseOverRenderArea)
-				{
-					distance = Mathf.Clamp(distance * Mathf.Exp(-Input.GetAxis("Mouse ScrollWheel") * zoomSpeed), distanceMin, distanceMax);
-				}
+			if (Input.GetMouseButton(0) || mouseOverRenderArea)
+			{
+				distance = Mathf.Clamp(distance * Mathf.Exp(-Input.GetAxis("Mouse ScrollWheel") * zoomSpeed), distanceMin, distanceMax);
+			}
 
-				Vector3 negDistance = new Vector3(0.0f, 0.0f, -distance);
-				Vector3 position = rotation * negDistance + target.position;
+			Vector3 negDistance = new Vector3(0.0f, 0.0f, -distance);
+			Vector3 position = rotation * negDistance + targetPosition;
 
-				transform.rotation = rotation;
-				transform.position = position;
-			}
+			transform.rotation = rotation;
+			transform.position = position;
+
+			prevMousePosition = Input.mousePosition;
+		}
+
+		private Vector3 ProjectScreenPointToTargetPlane(Vector3 screenPosition)
+		{
+			var ray = camera.ScreenPointToRay(screenPosition);
+			var planeDistance = distance / Mathf.Cos(Mathf.Deg2Rad * Vector3.Angle(camera.transform.forward, ray.direction));
```

Comment:
Do you need to handle divide by zero here?

Reply:
It's not possible for this to be zero with a perspective camera because the angles between the ray and the camera's forward can never quite reach 90 degrees.